### ISSUE TITLE
Why is resp.vary type inconsistent?

### ISSUE THREAD

**Issue by dorfsmay** [falconry/falcon#1253]
resp.vary setter expect a list but it's getter returns a string:

    resp.vary = 'blah'
    print(resp.vary) ==> 'b', 'l', 'a', 'h'
    print(type(resp.vary) ==> <class 'str'>

This is inconsistent and means that adding a header requires a string manipulation when it really should be as simple as `resp.vary.append('Origin')`.

Happy to create a PR for this, but it will break backaward incompatibility, maybe better for a major release.

**Comment by kgriffs** [comment]
Hi @dorfsmay, I'd love to get your help on this! It is indeed clunky and has actually been on my radar for some time. We can finally address it in 2.0, since we will can make breaking changes. 

Since there are a few different ways this behavior could be implemented, it would be best if you first outlined an approach and shared it here for feedback from some of the main project contributors.

Thanks!

**Comment by Nateyo** [comment]
I'm going to take a look into this.

**Comment by Nateyo** [comment]
So this affects a lot of the headers because it involves header_property. I think maybe it's odd behavior when setting headers to immediately call the normalizer when setting. I'm not quite sure how to structure it yet, but would it be an ok approach to normalizer the headers dictionary before calling headers = self._headers in _wsgi_headers?

**Comment by Nateyo** [comment]
I'm wondering if it's worth it to change how headers are handled in general. It appears there are all the set/get/append_header methods in response.py, but then we have these HeaderProperty properties that are available for what I assume are the more common headers(?) so that one can use response.vary vs. get_header("Vary")/append_header("Vary"). It seems having these two separate methodologies for accessing headers is inconsistent.

Perhaps headers should be some type of bare class (I'd say dataclass but that'd break python2) so accessing headers is more consistent (resp.headers.vary, vs. resp.vary). Maybe I don't know the response protocols well enough as to why there are a bunch of properties on Response itself that modify Response._headers.

**Comment by kgriffs** [comment]
I'm always open to new approaches, although we need to look carefully at (1) the performance impact on the new approach, (2) how the change in approach might impact the interface and semantics and make sure it is not too disruptive (if it is, we might need to try to shim old behavior of stagger the breaking changes across multiple releases), and (3) the maintenance burden of the new approach.

Anyway, we will need some way of storing the value as originally set so that it can be returned by the getter, and then finalizing that value to the raw string value for things like `resp.get_header()`, `resp.headers`, and `resp._wsgi_headers()`.

It would be great to have this in 2.0, but we can move it to 3.0 if necessary(which is on a shorter cycle than 2.0 was, FWIW) if it looks like it is going to take more than a couple of weeks to land.

**Comment by Nateyo** [comment]
@kgriffs I think this is probably unlikely for 2.0 release, mostly because headers are involved all over the place so the PR will be pretty big. I've been tinkering and I have some questions:

- How tied down are we to the idea of `HeaderProperty`? I think its a slight violation of DRY that we have header mutation functions and then separate `HeaderPropertys `which define their own get/set/etc, resulting in the above strange behavior.
- I think you mention that python2 will be deprecated for 3.0, is that correct?

I've been tinkering with a `Headers `class approach, a lot of other frameworks use something similar, but since falcon tries to be speedy, we could keep the headers on the response object. Although response.py does become somehwat easier to read considering 50% of it is HeaderProperty definitions and manipulation.

**Comment by kgriffs** [comment]
> I think this is probably unlikely for 2.0

Yes, and in fact we may have to push a few other things out as well in the interest of time. I'll go ahead and reassign the milestone.

>  How tied down are we to the idea of `HeaderProperty`?

The original idea of the header property was to provide a way to generate performant getters and setters without having to copy-paste code everywhere. However, things have indeed gotten a bit less elegant with the introduction of header manipulations on the values as they are set, vs. just settings the header values directly with the raw strings. If there is a more elegant way of approaching this that does not compromise performance, I'd be happy to entertain it.

> I think you mention that python2 will be deprecated for 3.0, is that correct?

Python 2 is actually going to be deprecated in the next release (2.0). We are going to completely drop support for anything below Python 3.5 in the Falcon 3.0 release.
